lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -0,0 +1,1538 @@
+///    modifying the port list of the module. This is done by first determining
+///    the name of the reset signal in the parent module, which is either the
+///    name of the port or wire declaration. We then look for an existing
+///    `asyncreset` port in the port list and reuse that as reset. If no port
+///    with that name was found, or the existing port is of the wrong type, a
+///    new port is inserted into the port list.
+///
+struct InferResetsPass : public InferResetsBase<InferResetsPass> {
+  void runOnOperation() override;
+  void runOnOperationInner();
+
+  // Copy creates a new empty pass (because ResetMap has no copy constructor).
+  using InferResetsBase::InferResetsBase;
+  InferResetsPass(const InferResetsPass &) {}
+
+  //===--------------------------------------------------------------------===//
+  // Reset type inference
+
+  LogicalResult traceResets(CircuitOp circuit);
+  void traceResets(InstanceOp inst);
+  void traceResets(Value dst, Value src, Location loc);
+  void traceResets(Value value);
+  void traceResets(FIRRTLType dstType, Value dst, unsigned dstID,
+                   FIRRTLType srcType, Value src, unsigned srcID, Location loc);
+
+  LogicalResult inferResets();
+  LogicalResult inferReset(ResetNet *net);
+
+  LogicalResult updateResets();
+  LogicalResult updateReset(ResetNet *net);
+  bool updateReset(FieldRef field, FIRRTLType resetType);
+
+  //===--------------------------------------------------------------------===//
+  // Async reset implementation
+
+  LogicalResult collectAnnos(CircuitOp circuit);
+  LogicalResult collectAnnos(FModuleOp module);
+
+  LogicalResult buildDomains(CircuitOp circuit);
+  void buildDomains(FModuleOp module, const InstancePath &instPath,
+                    Value parentReset, InstanceGraph &instGraph,
+                    unsigned indent = 0);
+
+  void determineImpl();
+  void determineImpl(FModuleOp module, ResetDomain &domain);
+
+  LogicalResult implementAsyncReset();
+  LogicalResult implementAsyncReset(FModuleOp module, ResetDomain &domain);
+  void implementAsyncReset(Operation *op, FModuleOp module, Value actualReset,
+                           SmallVectorImpl<Operation *> &deleteOps,
+                           SmallVectorImpl<std::pair<Value, Value>> &connects);
+
+  //===--------------------------------------------------------------------===//
+  // Analysis data
+
+  /// A map of all traced reset networks in the circuit.
+  ResetMap resetMap;
+
+  /// The annotated reset for a module. A null value indicates that the module
+  /// is explicitly annotated with `ignore`. Otherwise the port/wire/node
+  /// annotated as reset within the module is stored.
+  DenseMap<Operation *, Value> annotatedResets;
+
+  /// The reset domain for a module. In case of conflicting domain membership,
+  /// the vector for a module contains multiple elements.
+  MapVector<Operation *, SmallVector<std::pair<ResetDomain, InstancePath>, 1>>
+      domains;
```

Comment:
Lots of things are casting the key to FModuleOp.  Can you use FModuleOp instead of Operation*?

Reply:
Yes, changed that. This requires me to derive `DenseMapInfo<FModuleOp>` which is something that should definitely come out of TableGen I think. Probably a good change to make upstream.